### PR TITLE
Porting dplyr to c++11

### PR DESCRIPTION
**Pull request by jayhesselberth**
This PR illustrates a minimal set of changes needed to  dplyr c++ to enable C++11 compatibility. All-in-all this was surprisingly little work. @krlmlr and I have discussed the utility of the C++ side of dplyr (the API is very useful, but could be more consistent and stable).

This PR builds cleanly on my OS X box with `clang++`, and all tests pass. You will see warnings if you build against current Rcpp (0.12.7; concerning `##__VA_ARGS__`); these will go away in Rcpp [0.12.7.5](https://github.com/RcppCore/Rcpp/issues/576) There are issues with [`g++` on travis](https://travis-ci.org/jayhesselberth/dplyr); my guess is that these are trivial fixes (e.g. passing correct compiler flags), but someone else could weigh in.

Advantages:

* Removes the dependency on boost (`BH`).
* Future work on dplyr could use C++11 features (`auto` etc).

Disadvantages:

* Would prevent building of `dplyr` with older compilers that don't support the 11 standard.

Other than replacement of boost things  with the equivalent from c++11 `std` (e.g. (`shared/unique_ptrs`, `unordered_map/set`), there are a few other changes:

* `collapse` was renamed to `collapse_string` to prevent a clash with `Rcpp::sugar::collapse()`

* had to define `std::hash<Rcomplex>`. My implementation could probably be improved.

### CODE REVIEW

File: inst/include/tools/SymbolMap.h [hunk 1]
```diff
@@ -19,7 +19,7 @@ namespace dplyr {
 
   class SymbolMap {
   private:
-    dplyr_hash_map<SEXP, int> lookup;
```
Comment (krlmlr) [inline inst/include/tools/SymbolMap.h]: Would it be useful to keep the typedefs?

File: inst/include/tools/hash.h [hunk 2]
```diff
@@ -1,30 +1,29 @@
...
-    #define dplyr_hash_map boost::unordered_map
-  #endif
-#endif
+namespace std {
+  template <>
+  struct hash<Rcomplex>
+  {
+    size_t operator()(const Rcomplex & x) const
+    {
+		return hash<int>()(x.i);
```
Comment (krlmlr) [inline inst/include/tools/hash.h]: Is this supposed to call hash_value(Rcomplex&) ?

File: inst/include/tools/SymbolMap.h (hunk 1 above)
Comment (jayhesselberth) [inline inst/include/tools/SymbolMap.h]: As long as we use c++11, I think it's a clean transition to `std::unordered_map` from `dplyr_hash_map` (and to `std::unordered_set` from `dplyr_hash_set`).

File: inst/include/tools/hash.h (hunk 2 above)
Comment (jayhesselberth) [inline inst/include/tools/hash.h]: yep. done in 0a798a9.

File: inst/include/dplyr/main.h [hunk 3]
```diff
@@ -2,11 +2,12 @@
 #define dplyr_dplyr_main_H
 
 #include <Rcpp.h>
-#include <dplyr/workarounds/static_assert.h>
 #include <solaris/solaris.h>
 #include <dplyr/config.h>
 #include <dplyr/workarounds.h>
 
+#include <functional>
```
Comment (krlmlr) [inline inst/include/dplyr/main.h]: What's the benefit of having this include here?

File: inst/include/dplyr/visitor_set/VisitorSetHash.h [hunk 4]
```diff
@@ -16,7 +16,7 @@ namespace dplyr {
       }
       size_t seed = obj.get(0)->hash(j);
       for (int k=1; k<n; k++) {
-        boost::hash_combine(seed, obj.get(k)->hash(j));
+       hash_combine(seed, obj.get(k)->hash(j));
```
Comment (krlmlr) [inline inst/include/dplyr/visitor_set/VisitorSetHash.h]: Please make sure that the astyle test passes.

File: inst/include/tools/SymbolMap.h [hunk 5]
```diff
@@ -103,7 +103,7 @@ namespace dplyr {
         int idx = index.pos;
         names.erase(names.begin() + idx);
 
-        for (dplyr_hash_map<SEXP, int>::iterator it=lookup.begin(); it != lookup.end();) {
+        for (std::unordered_map<SEXP, int>::iterator it=lookup.begin(); it != lookup.end();) {
```
Comment (krlmlr) [inline inst/include/tools/SymbolMap.h]: Can we now use `auto` or a range-based `for` loop?

File: inst/include/tools/hash.h [hunk 6]
```diff
@@ -1,31 +1,30 @@
...
-  #if defined(_WIN32)
-    #define dplyr_hash_set RCPP_UNORDERED_SET
-  #else
-    #include <boost/unordered_set.hpp>
-    #define dplyr_hash_set boost::unordered_set
-  #endif
-#endif
+// http://stackoverflow.com/questions/2590677/how-do-i-combine-hash-values-in-c0x
+template <class T>
+inline void hash_combine(std::size_t& seed, const T& v)
```
Comment (krlmlr) [inline inst/include/tools/hash.h]: I'd prefer a pure function here.

File: inst/include/dplyr/main.h (hunk 3 above)
Comment (jayhesselberth) [inline inst/include/dplyr/main.h]: No benefit for this file per se, but need it somewhere for `std::hash`. Could move it down in the headers.

File: inst/include/tools/SymbolMap.h (hunk 5 above)
Comment (jayhesselberth) [inline inst/include/tools/SymbolMap.h]: we could indeed. i held off using `auto` for this PR but could have a once through and make all those changes in a separate commit.

File: inst/include/tools/hash.h (hunk 6 above)
Comment (jayhesselberth) [inline inst/include/tools/hash.h]: will do.

**Comment by kevinushey** [comment]
Don't we want to ensure `dplyr` adheres to the C++98 standard so it can build on older linux systems (e.g. RHEL5)?

**Comment by jayhesselberth** [comment]
Absolutely. Is `dplyr11` a better way forward? Using `Rcpp11`?

**Comment by kevinushey** [comment]
I think a port of `dplyr` to C++11 would be great (we don't necessarily need a whole new package), but now isn't quite the time -- IMHO it would be good to wait a bit longer until C++11 support is ubiquitous even in enterprise environments.

(of course @hadley has the last word here!)

**Comment by krlmlr** [comment]
Sorry: I didn't expect that the move to C++11 could cause compatibility issues.

It looks like we still can make use of most changes in this PR, so that switching to C++11 is achieved just by changing a `#define`. We could even add a Travis test that checks if the package compiles in "C++11 mode".

**Comment by romainfrancois** [comment]
FWIW, I'm planning a complete rework of `Rcpp11` with a different set of goals. 
https://github.com/rstats-cpp

**Comment by krlmlr** [comment]
@hadley: How important is C++0x support for dplyr?

**Comment by hadley** [comment]
It's still needed for RHEL, which is a platform that RStudio supports. We can use C++11 for smaller, more experimental packages, but big things like dplyr need to stay with C++03.

**Comment by krlmlr** [comment]
Thanks. @jayhesselberth: Would you like to work on transforming this PR so that C++11 mode can be turned on with a `#define`? This will make the code "C++11-ready", and usable for `LinkingTo:` other C++11 packages.

**Comment by hadley** [comment]
@krlmlr At this point, I'd rather not have to maintain two separate code paths. Lets hold off on this for six months or so, and check back with what's happening in RHEL-land.

**Comment by krlmlr** [comment]
Let's keep this open as a reference, we should be able to replicate the necessary steps later. Thanks anyway @jayhesselberth, sorry for the extra work my ignorance has caused.

**Comment by hadley** [comment]
Four months later, I'd say it's not looking likely we'll be able to use C++11 in dplyr. RHEL is important for RStudio, and C++11 support on commonly available versions is still not good enough.
